ns for fx: support comparing fp32 vs fp32_prepared/int8, except shadowed

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61380
* #61377
* #61360
* #61359
* #61358
* #61357
* #61323
* **#61129**

Summary:

Adds support the comparing fp32 model (without quantization) to a
fp32 model prepared with quantization or a quantized model. The main missing feature was
handling conv-bn fusion, since this fusion for PTQ happens outside
of quantization patterns.

Adds testing for this case for comparing weights and comparing
activations

Adds a TODO for also handling this for shadow activations, we need to
first stop removing observers in graph passes before we can add
this support, will be in a future PR.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2_qat
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels.test_compare_activations_conv
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29520009](https://our.internmc.facebook.com/intern/diff/D29520009)

vkuzo committed:
ns for fx: support comparing fp32 vs fp32_prepared, except shadowed

Summary:

Adds support the comparing fp32 model (without quantization) to a
fp32 model prepared with quantization. The main missing feature was
handling conv-bn fusion, since this fusion for PTQ happens outside
of quantization patterns.

Adds testing for this case for comparing weights and comparing
activations

Adds a TODO for also handling this for shadow activations, we need to
first stop removing observers in graph passes before we can add
this support, will be in a future PR.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2_qat
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels.test_compare_activations_conv
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).

vkuzo committed:
Update on "ns for fx: support comparing fp32 vs fp32_prepared, except shadowed"


Summary:

Adds support the comparing fp32 model (without quantization) to a
fp32 model prepared with quantization. The main missing feature was
handling conv-bn fusion, since this fusion for PTQ happens outside
of quantization patterns.

Adds testing for this case for comparing weights and comparing
activations

Adds a TODO for also handling this for shadow activations, we need to
first stop removing observers in graph passes before we can add
this support, will be in a future PR.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2_qat
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels.test_compare_activations_conv
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29520009](https://our.internmc.facebook.com/intern/diff/D29520009)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).

raghuramank100 commented on test/quantization/fx/test_numeric_suite_fx.py:
```diff
@@ -758,19 +762,21 @@ def _test_extract_weights(
-            self.assert_ns_compare_dict_valid(results)
+            # test both m vs mp and mp vs mq
+            for m1, m2 in ((m, mp), (mp, mq)):
+                results = extract_weights_fun('a', mp, 'b', mq)
```
Should this be m1 and m2?

vkuzo replied:
yes, great catch, will fix that

vkuzo replied:
yes, great catch, will fix that

hx89 reviewed (approved):
lgtm! Just wonder what would be the use case for comparing fp32 and fp32_prepared? Is it for sanity check to make sure fp32 is numerically the same after prepare and helping debug the implementation of prepare function?

vkuzo commented:
> lgtm! Just wonder what would be the use case for comparing fp32 and fp32_prepared? Is it for sanity check to make sure fp32 is numerically the same after prepare and helping debug the implementation of prepare function?

there was a use case discussed during design review to compare fp32 to quantized.  This PR implements this functionality (once you can compare A to B and B to C, you can compare A to C).  Let me update the summary to say this.

vkuzo committed:
Update on "ns for fx: support comparing fp32 vs fp32_prepared/int8, except shadowed"


Summary:

Adds support the comparing fp32 model (without quantization) to a
fp32 model prepared with quantization or a quantized model. The main missing feature was
handling conv-bn fusion, since this fusion for PTQ happens outside
of quantization patterns.

Adds testing for this case for comparing weights and comparing
activations

Adds a TODO for also handling this for shadow activations, we need to
first stop removing observers in graph passes before we can add
this support, will be in a future PR.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2_qat
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels.test_compare_activations_conv
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29520009](https://our.internmc.facebook.com/intern/diff/D29520009)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).

vkuzo committed:
Update on "ns for fx: support comparing fp32 vs fp32_prepared/int8, except shadowed"


Summary:

Adds support the comparing fp32 model (without quantization) to a
fp32 model prepared with quantization or a quantized model. The main missing feature was
handling conv-bn fusion, since this fusion for PTQ happens outside
of quantization patterns.

Adds testing for this case for comparing weights and comparing
activations

Adds a TODO for also handling this for shadow activations, we need to
first stop removing observers in graph passes before we can add
this support, will be in a future PR.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2_qat
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels.test_compare_activations_conv
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29520009](https://our.internmc.facebook.com/intern/diff/D29520009)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).
